Make to_tensor and normalize to accept 3D or 4D tensor inputs

sandeep-krishnamurthy opened the pull request:
## Description ##
With this change:
- `ToTensor` transformation operator can take 3D (h, w, c) or 4D (n, h, w, c) at once.
- `Normalize` transformation operator can take 3D (c, h, w) or 4D (n, c, h, w) at once.
- Also parallelized the operator with omp
- This change will be required to fuse transformation pipeline with network graph during inference. Where inputs are usually 4D inputs (n, c, h, w).
- This is backward compatible change, this do not change any existing behavior.

In another PR, I will work on to support list of 3D tensors (list of images) as input for different shape image tensors use case.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Support 3D or 4D input tensors to to_tensor and normalize transformation operator.

### TODO ###
- API documentation updates. (WIP)

@stu1130 @zhreshold @apeforest

stu1130 commented on src/operator/image/image_random-inl.h:
```diff
@@ -123,28 +159,50 @@ inline bool NormalizeShape(const nnvm::NodeAttrs& attrs,
+
+      for (int i = 0; i < channel; ++i) {
+        DType mean = param.mean[param.mean.ndim() > 1 ? i : 0];
+        DType std_dev = param.std[param.std.ndim() > 1 ? i : 0];
```
mean and std_dev should be float type defined by line 115, 116

stu1130 commented on src/operator/image/image_random-inl.h:
```diff
@@ -62,6 +67,23 @@ inline bool ToTensorType(const nnvm::NodeAttrs& attrs,
+  MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+      float* output = outputs[0].dptr<float>();
+      DType* input = inputs[0].dptr<DType>();
+
```
```suggestion
#pragma omp parallel for collapse(2)
```
would this make better performance?

sandeep-krishnamurthy replied:
Cannot be used inside Macro.

stu1130 replied:
How about checking the DType by ourself without using the Macro

stu1130 commented on src/operator/image/image_random-inl.h:
```diff
@@ -123,28 +159,50 @@ inline bool NormalizeShape(const nnvm::NodeAttrs& attrs,
+        DType mean = param.mean[param.mean.ndim() > 1 ? i : 0];
+        DType std_dev = param.std[param.std.ndim() > 1 ? i : 0];
+        for (int j = 0; j < length; ++j) {
+          output[step + i*length + j] = (input[step + i*length + j] - mean) / std_dev;
```
if input is int, should it be int or float after nomarlization? I prefer float here

sandeep-krishnamurthy replied:
Good point, it should be float. Making the change.

stu1130 commented on src/operator/image/image_random-inl.h:
```diff
@@ -123,28 +159,50 @@ inline bool NormalizeShape(const nnvm::NodeAttrs& attrs,
+    const int batch_size = inputs[0].shape_[0];
+    const int length = inputs[0].shape_[2] * inputs[0].shape_[3];
+    const int channel = inputs[0].shape_[1];
+    const int step = channel*length;
```
nit: 
```suggestion
    const int step = channel * length;
```

sandeep-krishnamurthy replied:
Done

stu1130 commented on tests/python/unittest/test_gluon_data_vision.py:
```diff
@@ -19,30 +19,66 @@
 @with_seed()
 def test_to_tensor():
+    # 3D Input
     data_in = np.random.uniform(0, 255, (300, 300, 3)).astype(dtype=np.uint8)
```
```suggestion
    data_in = nd.random.uniform(0, 255, (300, 300, 3)).astype(dtype=np.uint8)
```
directly initialize ndarray would be better?

roywei commented:
@mxnet-label-bot add[Gluon, Data-loading, Operator]

apeforest commented on python/mxnet/gluon/data/vision/transforms.py:
```diff
@@ -154,10 +157,29 @@ class Normalize(HybridBlock):
+    Examples
+    --------
+    >>> transformer = transforms.Normalize(mean=(0, 1, 2), std=(3, 2, 1))
+    >>> image = mx.nd.random.uniform(0, 1, (3, 4, 2))
```
Can you give an example of 4D (N x C x H x W) here?

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -47,9 +47,14 @@ inline bool ToTensorShape(const nnvm::NodeAttrs& attrs,
+      << "Input image must have shape (height, width, channels), or "
+      << "(N, height, width, channels) but got " << shp;
+  if (shp.ndim() == 3) {
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape({shp[2], shp[0], shp[1]}));
```
It would be clearer to define enum constant N, W, C, H instead of using 0, 1, 2, 3

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -62,6 +67,23 @@ inline bool ToTensorType(const nnvm::NodeAttrs& attrs,
+
+      for (int l = 0; l < length; ++l) {
+        for (int c = 0; c < channel; ++c) {
+          output[step + c*length + l] = static_cast<float>(input[step + l*channel + c]) / 255.0f;
```
255.0f is already a float, so this cast may not be needed here.

apeforest replied:
Another question is why 255.0f? Maybe using a constant variable with clear name is more readable.

sandeep-krishnamurthy replied:
Agreed. Making the change.

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -70,19 +92,23 @@ void ToTensor(const nnvm::NodeAttrs &attrs,
+
+    #pragma omp parallel for
+    for (auto n = 0; n < batch_size; ++n) {
+      ToTensorImpl(inputs, outputs, length, channel, n*step);
```
Good change!

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -103,14 +129,24 @@ inline bool NormalizeShape(const nnvm::NodeAttrs& attrs,
+      << "Input tensor must have shape (channels, height, width), or "
+      << "(N, channels, height, width), but got " << dshape;
+
+  uint32_t nchannels;
```
Use int or int32_t. See: https://google.github.io/styleguide/cppguide.html#Integer_Types

sandeep-krishnamurthy replied:
Thanks. Good learning!

apeforest commented on tests/python/unittest/test_gluon_data_vision.py:
```diff
@@ -19,30 +19,66 @@
+    # 3D Input
     data_in = np.random.uniform(0, 255, (300, 300, 3)).astype(dtype=np.uint8)
-    out_nd = transforms.ToTensor()(nd.array(data_in, dtype='uint8'))
+    out_nd = transforms.ToTensor()(nd.array(data_in))
```
Why remove the dtype in original 3D input test?

apeforest commented on tests/python/unittest/test_gluon_data_vision.py:
```diff
@@ -19,30 +19,66 @@
+    assert_almost_equal(data_expected_4d, out_nd_4d.asnumpy())
+
+    # Invalid Input - Neither 3D or 4D input
+    invalid_data_in = nd.random.uniform(0, 1, (5, 5, 3, 300, 300)).astype(dtype=np.float32)
```
👍

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -70,19 +92,23 @@ void ToTensor(const nnvm::NodeAttrs &attrs,
+    const int batch_size = inputs[0].shape_[0];
+    const int length = inputs[0].shape_[1] * inputs[0].shape_[2];
+    const int channel = inputs[0].shape_[3];
+    const int step = channel * length;
```
How large can this value be in practice?

stu1130 replied:
Good point, IMO make step unsigned long long int would be enough?

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -70,19 +92,23 @@ void ToTensor(const nnvm::NodeAttrs &attrs,
+    ToTensorImpl(inputs, outputs, length, channel);
+  } else if (inputs[0].ndim() == 4) {
+    // 4D input batch of images
+    const int batch_size = inputs[0].shape_[0];
```
How large can this value be in practice?

apeforest commented on src/operator/image/image_random-inl.h:
```diff
@@ -47,9 +47,14 @@ inline bool ToTensorShape(const nnvm::NodeAttrs& attrs,
+  if (shp.ndim() == 3) {
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape({shp[2], shp[0], shp[1]}));
+  } else if (shp.ndim() == 4) {
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, TShape({shp[0], shp[3], shp[1], shp[2]}));
```
Can shp[0] be zero?

sandeep-krishnamurthy commented:
@stu1130 @apeforest - I have addressed your comments. I have raised a separate PR - https://github.com/apache/incubator-mxnet/pull/13802 only for Normalize. It also includes GPU support.

Closing this PR. I request you to review other PR. Thanks.
